### ISSUE TITLE
Wrong random code generated

### ISSUE THREAD

**Issue by kdomagal** [androidannotations/androidannotations#2125]
I have a very strange issue with android annotations and android studio. It seems like it assigns on click events to random elements instead of those it was annotated.  As you can see below click events were assigned to completely different element ids in generated code. I think it happens only when the code is generated in android studio. When I compile it with gradle in command line the generated code is ok. Not sure if kotlin has something to do with it

**AndroidAnnotations version**:
4.4.0
**Android compile SDK version**:
26
**Annotated code**:
```kotlin
@Click(R.id.button_logout)
    fun logout() {
        progressBar.visibility = VISIBLE
        button_logout.isEnabled = false
        loginService.logout()
    }

    @Click(R.id.profileImage)
    fun refreshProfile() {
        profileService.refreshCachedProfile()
    }
 @Click(R.id.marketExploreButton)
    fun marketExplore() {
        val intent = Intent(baseContext, MarketExploreActivity_::class.java)
        startActivity(intent)
    }

    @Click(R.id.savedPropertiesButton)
    fun checkSavedProperties() {
        val intent = Intent(baseContext, SavedPropertiesActivity_::class.java)
        startActivity(intent)
    }

    @Click(R.id.investmentsButton)
    fun checkInvestments(){
        val intent = Intent(baseContext, InvestmentListActivity_::class.java)
        startActivity(intent)
    }

    @Click(R.id.timeLineButton)
    fun displayTimeLine(){
        val intent = Intent(baseContext, TimeLineActivity_::class.java)
        startActivity(intent)
    }
```
**Actual generated code**:
```java
   View view_end_padder = hasViews.internalFindViewById(R.id.end_padder);
        View view_right_icon = hasViews.internalFindViewById(R.id.right_icon);
        View view_daysPassedText = hasViews.internalFindViewById(R.id.daysPassedText);
        View view_dayesTextView = hasViews.internalFindViewById(R.id.dayesTextView);
        View view_jobSalaryText = hasViews.internalFindViewById(R.id.jobSalaryText);
        View view_investmentsButton = hasViews.internalFindViewById(R.id.investmentsButton);

        if (view_end_padder!= null) {
            view_end_padder.setOnClickListener(new OnClickListener() {

                @Override
                public void onClick(View view) {
                    ProfileActivity_.this.logout();
                }
            }
            );
        }
        if (view_right_icon!= null) {
            view_right_icon.setOnClickListener(new OnClickListener() {

                @Override
                public void onClick(View view) {
                    ProfileActivity_.this.refreshProfile();
                }
            }
            );
        }
        if (view_daysPassedText!= null) {
            view_daysPassedText.setOnClickListener(new OnClickListener() {

                @Override
                public void onClick(View view) {
                    ProfileActivity_.this.marketExplore();
                }
            }
            );
        }
        if (view_dayesTextView!= null) {
            view_dayesTextView.setOnClickListener(new OnClickListener() {

                @Override
                public void onClick(View view) {
                    ProfileActivity_.this.checkSavedProperties();
                }
            }
            );
        }
        if (view_jobSalaryText!= null) {
            view_jobSalaryText.setOnClickListener(new OnClickListener() {

                @Override
                public void onClick(View view) {
                    ProfileActivity_.this.checkInvestments();
                }
            }
            );
        }
        if (view_investmentsButton!= null) {
            view_investmentsButton.setOnClickListener(new OnClickListener() {

                @Override
                public void onClick(View view) {
                    ProfileActivity_.this.displayTimeLine();
                }
            }
            );
        }
```

**Comment by mortenholmgaard** [comment]
I do experience the exact same thing with Kotlin. When I do a rebuild it starts working again, but only for some time - don't know what is triggering the wrong generated code.

**Comment by mortenholmgaard** [comment]
I don't know but could it be related to something about instant run, and the way it compiles there is somewhat different from the normal run case?

**Comment by dodgex** [comment]
Sorry for the late response, could you please provide a small sample project where we can reproduce the issue? Thanks!

**Comment by kdomagal** [comment]
I'm using:
Android Studio 3.1.2
Build #AI-173.4720617, built on April 13, 2018
JRE: 1.8.0_152-release-1024-b01 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Mac OS X 10.13.4

AndroidAnnotations version 4.4.0
sdk version 26
kotlin 1.2.41

'com.android.support.constraint:constraint-layout:1.1.0' if that has something do with it. 

Just create a simple app with the above versions and a simple layout and @click annotations.. happens almost every time. Maybe it's some bug in Android Studio..

**Comment by dodgex** [comment]
I'm unable to reproduce this with the following setup:
```
Android Studio 3.1.3
Build #AI-173.4819257, built on June 4, 2018
JRE: 1.8.0_152-release-1024-b02 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Windows 10
```

```
// Top-level build file where you can add configuration options common to all sub-projects/modules.

buildscript {
    ext.kotlin_version = '1.2.41'
    repositories {
        google()
        jcenter()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.1.2' // and 3.1.2
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        google()
        jcenter()
    }
}

task clean(type: Delete) {
    delete rootProject.buildDir
}
```

```
apply plugin: 'com.android.application'
apply plugin: 'kotlin-kapt'
apply plugin: 'kotlin-android'
apply plugin: 'kotlin-android-extensions'

android {
    compileSdkVersion 26
    defaultConfig {
        applicationId "eu.sourceway.aakotlinbugtest"
        minSdkVersion 21
        targetSdkVersion 26
        versionCode 1
        versionName "1.0"
        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
}

dependencies {
    implementation"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    implementation 'com.android.support:appcompat-v7:26.0.0'
    implementation 'com.android.support.constraint:constraint-layout:1.1.0'
    implementation 'com.android.support:design:26.0.0'

    kapt  "org.androidannotations:androidannotations:4.4.0"
    implementation "org.androidannotations:androidannotations-api:4.4.0"
}
```
or (latest versions)
```
apply plugin: 'com.android.application'

apply plugin: 'kotlin-kapt'

apply plugin: 'kotlin-android'

apply plugin: 'kotlin-android-extensions'

android {
    compileSdkVersion 28
    defaultConfig {
        applicationId "eu.sourceway.aakotlinbugtest"
        minSdkVersion 21
        targetSdkVersion 28
        versionCode 1
        versionName "1.0"
        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
}

dependencies {
    implementation"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    implementation 'com.android.support:appcompat-v7:28.0.0-alpha3'
    implementation 'com.android.support.constraint:constraint-layout:1.1.1'
    implementation 'com.android.support:design:28.0.0-alpha3'

    kapt  "org.androidannotations:androidannotations:4.4.0"
    implementation "org.androidannotations:androidannotations-api:4.4.0"
}

```

**Comment by dodgex** [comment]
If possible could you *please* provide a minimal project that can reproduce the issue? Thanks!

**Comment by xinaiz** [comment]
I think I managed to reproduce the bug. In my case it happens when using AA with butterknife, but it looks like ids are somehow moved by offest, and butterknife fails are simply much more painful. I will post project when I'm home. Should I post files here? (5 files) or do you prefer whole project? 

TL;DR In short, ButterKnife`s @OnClick in ViewHolder in Adapter tried to click button that was defined above recyclerView, thus these ids are defined really close.
`java.lang.IllegalStateException: Required view 'randomText' with ID 2131165281 for method 'randomButtonInRecycler' was not found. If this view is optional add '@Nullable' (fields) or '@Optional' (methods) annotation.`
The annotation in ViewHolder was:
`@OnClick(R.id.randomButtonInRecycler)`
 anyways, will post related files later in desired format.

**Comment by dodgex** [comment]
Would be nice to have a full project either as git repo (preferred) or zip. so that i just need to run `gradlew build` to see the problem.

**Comment by mortenholmgaard** [comment]
It seems to have something to do with layout files - adding and removing (and modify?) layout files. When I do that it happens a lot more then else. I have not been able to narrow it more down.
Does this ring a bell for you @dodgex - something listening for something like that?

**Comment by dodgex** [comment]
So you mean when you edit a `layout.xml` file? I have no concrete idea but maybe it is somehow involved with how we extract view ids while processing the annotations.

but i have no concrete idea what could cause this behaviour. especially as it seems to only be kotlin related.

**Comment by mortenholmgaard** [comment]
I happens when I remove/add/edit a view in a layout.xml file, yes. It might be related to kotlin-android-extensions and the findViewById shortcuts it has - just an idea.

**Comment by dodgex** [comment]
Does this also happen if you use the `resName` attribute instead of passing the id?

**Comment by kdomagal** [comment]
I think `resName` fixes the problem

**Comment by kdomagal** [comment]
yes, I added `resName` attribute and the issue still happens

**Comment by dodgex** [comment]
More or less good news: I just got to reproduce this issue.

Repro case:
add views with ids `view0`, `view1`, `view2`, `view3`, `view4`. add click functions for 1-3. compile (first time) - everything ok
`R.id` values
```
    public static final int view0=0x7f07008f;
    public static final int view1=0x7f070090;
    public static final int view2=0x7f070091;
    public static final int view3=0x7f070092;
    public static final int view4=0x7f070093;
```
then remove  `view0`  and compile again. the click hander move up by one eg `view1` method is attached to `view2`
`R.id` values
```
    public static final int view1=0x7f07008f;
    public static final int view2=0x7f070090;
    public static final int view3=0x7f070091;
    public static final int view4=0x7f070092;
```

when you do a change to your activity and recompile it is correct again.

AND when you then add the `view0` again it moves the handlers again but the handler for `view1` gets assigned to `view0`

----

it looks like the resolving of the id that happens in android annotations does not get the "new" `R` class values and as a result resolves wrong elements

I'm afraid that this is a more complex issue. Need to check how AA resolves the ids...

**Comment by mortenholmgaard** [comment]
Any news on this?

**Comment by mortenholmgaard** [comment]
Have this been solved jet? 
It is starting to be an argument for me to ditch this library and find something else when using Kotlin..

**Comment by mortenholmgaard** [comment]
@WonderCsabo @dodgex Please have a look at this - it is critical for the future off this lib with Kotlin!

**Comment by dodgex** [comment]
Hey sorry for the delay, in the last time I didn't had much time and also currently not working on any Android related projects so it was hard to find time to work on this.

I just realized that one can use `resName` attribute from the annotation and from my tests it looks like this is a valid workaround for the given issue.

**Comment by WonderCsabo** [comment]
Hi guys,

Sorry for the delay again. I also debugged the problem, and we agree with @dodgex that the bug is not inside AndroidAnnotations. For some reason, wrong value is passed to the processor from the compiler plugin. I reported this [bug](https://youtrack.jetbrains.net/issue/KT-27487) in the Kotlin issue tracker. Please check it for more details, and upvote it.

As a workaround, you can do two things:

* Use the `resName` parameter in the annotation
* Do not pass any argument to the annotations, and use the id as the method name, like : `view1Clicked()`

**Comment by dodgex** [comment]
According to JetBrains issue tracker the issue should be fixed in latest kotlin dev version (1.3.30-dev-156) https://youtrack.jetbrains.com/issue/KT-27487

could someone test that? :)

**Comment by WonderCsabo** [comment]
Closing this issue, since the root problem was fixed in Kotlin 1.3.20.
